### PR TITLE
Added functions to convert .radis to JSON format and replicated all functions for JSON

### PR DESCRIPTION
**Pull request by anandxkumar**
#### Added function `convertRadisToJSON()`

- Loads `~/.radis` file using `getConfig()`
- Stores the configutation in JSON format - 
 
```
  {
  "HITEMP-OH": {
    "info": "HITEMP OH lines (0.0-43408.8 cm-1) with TIPS-2017 (through HAPI) for partition functions",
    "path": "/home/pipebomb/.radisdb/OH-13_HITEMP2020.h5",
    "format": "hitemp-radisdb",
    "parfuncfmt": "hapi",
    "download_url": "https://hitran.org/hitemp/data/bzip2format/13_HITEMP2020.par.bz2",
    "download_date": "12 Mar 2021",
    "wavenumber_min": "0.002967",
    "wavenumber_max": "43408.75294"
  }}
```
- On successful execution of function, json  data is dumped  in  `~/.radisjson` file.

#### Added function `getConfigJSON()`
- Reads the `~/.radisjson`  config file and returns it .

#### Added function `getDatabankListJSON()`
- Similar to `getDatabankList()` function, i.e. gets all databanks available but in `~/.radisjson` (JSON formatted config file).
- Using for test purpose to demonstrate config functions implementation using `~/.radisjson`.

### Further Discussion
- Do we need to create new alternate functions to implement all the functions in `radis/radis/misc/config.py` using `~/.radisjson` , then we can rename the functions to transit from configparser to JSON format?
- Other files needed to be looked other than config.py ?
- Further improvements in code and approach ?

Fixes : #167

### CODE REVIEW

**Comment by anandxkumar** [comment]
Hello @erwanp, can you please help me understand the process for transiting from configparser to JSON, is the approach I took above appropriate i.e. creating analog functions for each function in config.py dealing with creating, adding and extracting directly from a JSON type file?

File: radis/misc/config.py [hunk 1]
```diff
@@ -473,6 +494,83 @@ def printDatabankList():
...
+    Example
+    -------
+    original ~/.radis file format
+
+    [HITRAN-CO2-TEST]
+    info = HITRAN 2016 database, CO2, 1 main isotope (CO2-626), bandhead: 2380-2398 cm-1 (4165-4200 nm)
+    path = PATH_TO\radis\radis\test\files\hitran_co2_626_bandhead_4165_4200nm.par
+    format = hitran
+    parfuncfmt = hapi
+    levelsfmt = radis
```
Comment (erwanp) [inline radis/misc/config.py:509]:
```suggestion
    original ~/.radis file format::

        [HITRAN-CO2-TEST]
        info = HITRAN 2016 database, CO2, 1 main isotope (CO2-626), bandhead: 2380-2398 cm-1 (4165-4200 nm)
        path = PATH_TO\radis\radis\test\files\hitran_co2_626_bandhead_4165_4200nm.par
        format = hitran
        parfuncfmt = hapi
        levelsfmt = radis
```

File: radis/misc/config.py [hunk 2]
```diff
@@ -473,6 +494,83 @@ def printDatabankList():
...
+
+    -----------
+    Converted ~/.radisjson file format
+
+    {"HITRAN-CO2-TEST": {"info": "HITRAN 2016 database, CO2, 1 main isotope (CO2-626), bandhead: 2380-2398 cm-1 (4165-4200 nm)",
+    "path": "PATH_TO\\radis\\radis\\test\\files\\hitran_co2_626_bandhead_4165_4200nm.par",
+    "format": "hitran",
+    "parfuncfmt": "hapi",
+    "levelsfmt": "radis"}}
```
Comment (erwanp) [inline radis/misc/config.py:518]:
```suggestion
    Converted ~/.radisjson file format::

        {"HITRAN-CO2-TEST": {"info": "HITRAN 2016 database, CO2, 1 main isotope (CO2-626), bandhead: 2380-2398 cm-1 (4165-4200 nm)",
        "path": "PATH_TO\\radis\\radis\\test\\files\\hitran_co2_626_bandhead_4165_4200nm.par",
        "format": "hitran",
        "parfuncfmt": "hapi",
        "levelsfmt": "radis"}}
```

**Review by erwanp — CHANGES_REQUESTED** [review]
Hello @anandxkumar , nice work ! 
The approach of replicating the existing configparser functions is good. 

Few suggestions : 
- can we rename the .radisjson as `radis.json`  ? On Windows environnement .radis files are usually hidden, which creates issues for some users.
- can we add all the content in a main `databases` key   ? 
ie.
```
{"databases":{
{"HITRAN-CO2-TEST": {"info": "HITRAN 2016 database, CO2, 1 main isotope (CO2-626), bandhead: 2380-2398 cm-1 (4165-4200 nm)",
    "path": "PATH_TO\\radis\\radis\\test\\files\\hitran_co2_626_bandhead_4165_4200nm.par",
    "format": "hitran",
    "parfuncfmt": "hapi",
    "levelsfmt": "radis"}}
}
```
This will allow to use the same .json config file for other global RADIS parameters, such as printing/plotting references, etc. 



We need to add a test, to make sure existing .radis files are properly converted. Test can be added in [test_config.py](https://github.com/radis/radis/blob/develop/radis/test/misc/test_config.py)

What about the following (just an idea) : 
- rename the existing [setup_line_database](https://github.com/radis/radis/blob/develop/radis/test/utils.py#L207)  as `setup_line_database_configformat`
- write a new `setup_line_database` with the new json functions 
- in the test, call the `setup_line_database_configformat` to generate a .radis file, convert it with your Convert function, and ensure the result (the file directly, or the databases read from the files) are the same as when generated by `setup_line_database`  

?

**Comment by anandxkumar** [comment]
Hello @erwanp, some doubts to clarify- 
- What should be the proper format for multiple paths in JSON?
 ```
[MY-HITEMP-CO2]                                
info = HITEMP 2010 databank     
path =                          
       D:\Databases\HITEMP-CO2\hitemp_07
       D:\Databases\HITEMP-CO2\hitemp_08
       D:\Databases\HITEMP-CO2\hitemp_09
```
TO -> 
```
{"MY-HITEMP-CO2": 
        { "info":"HITEMP 2010 databank",
          "path":["D:\Databases\HITEMP-CO2\hitemp_07","D:\Databases\HITEMP-CO2\hitemp_08","D:\Databases\HITEMP-CO2\hitemp_09"]
}
```
So in function `addDatabankEntries(dbname, dict_entries)`, if there is a single path in dict_entries, we add that path like - `"path":"PATH/TO/HITEMP/CO2/hitemp_co2.par"` and if it's a list of paths then - `"path": ["path1","path2"]`?
In the function `getDatabankEntries(dbname, get_extra_keys=[])` the path is returned in form of a list even if it is a single path, so should we store path in list form even if there is a single path?

Also can you help me explain the suggestion you gave above - can we add all the content in a main `databases` key ?
Does this mean for every database index we will add all it's information? This thing is done if we execute a test or function (like setup_test_line_databases()), which calls the `addDatabankEntries(dbname, dict_entries)`, so did you mean to update this function to add directly to JSON file?


Clarify the workflow 
- Convert Function to convert existing .radis to radis.json
- Update all functions in config.py to work with radis.json instead of  .radis
- Add tests to check the working of convert function
- Modify existing tests for the all the functions in config.py

**Comment by erwanp** [comment]
Paths as a List within the Json dictionary is perfect. 
If it's a single path, allow it to be a string. 

For the `databases` key, my comment above didn't render properly. I fixed it, and copies it below. What i'm thinking of is : 

```
{"databases":{
{"HITRAN-CO2-TEST": {"info": "HITRAN 2016 database, CO2, 1 main isotope (CO2-626), bandhead: 2380-2398 cm-1 (4165-4200 nm)",
    "path": "PATH_TO\\radis\\radis\\test\\files\\hitran_co2_626_bandhead_4165_4200nm.par",
    "format": "hitran",
    "parfuncfmt": "hapi",
    "levelsfmt": "radis"}}
}
```

so we can have : 
```
{"databases":
   { "database1":... ,
    "database2": ... },
   }
# but also : 
  "other global radis parameters": ... ,
  "and user params" : ... 
  " etc. " 
}
```

**Comment by anandxkumar** [comment]
Hello @erwanp  , if we want to add all databases content to a databases key then what will be our identification criteria for databases? Like for now I believe that only databases informations are added into .radis so while converting it I can add it all the info like the below -
```
{"databases":
   { "database1": {"info": ....} ,
    "database2": ... },
   }
 }
 ```
 So later on other functions can be used to add global parameters to the JSON file.
 ```
{
# params: 
  "other global radis parameters": ... ,
  "and user params" : ... 
  " etc. " 
}
```
Also I have solution for the test, we can add a test in `test_config.py`, we will pass a flag parameter in the convert function and while calling the test we first use `setup_line_database_configformat`  with the flag that will create a test config file suppose `.radistest `and then convert it to save as `radistest.json`, then we use the `setup_line_database`  to create `radistest2.json` , and load both JSONs in variables and then delete the files we created (So nothing additional is being added) and check whether both variables are similar or not.
One last thing, suppose everything works fine, then we want to make sure every function and test uses the JSON functions, so we can replace the original functions with the JSON functionality (So the function name remains the same) and then rename the original functions as `function_name+_configformat` and use those for using executing the test we created?

Also can you share your `.radis `file? That will be a good test case.

File: radis/spectrum/spectrum.py [hunk 3]
```diff
@@ -98,7 +98,7 @@ class Spectrum(object):
         and wavenum is in cm-1
         example::
 
-        >>> {'radiance_noslit:':(wavenum, radiance_noslit),
+        >>> {'radiance_noslit':(wavenum, radiance_noslit),
```
Comment (erwanp) [inline radis/spectrum/spectrum.py:101]: nice catch @anandxkumar

**Comment by anandxkumar** [comment]
Yes @erwanp you are right, It was a typo. But I have removed the conditions for both `parfuncfmt` and `format`, and sticking with the original code as discussed over slack. 
Will we replace the JSON functions with the original function name so that the updated functions will  be used for all the test and functionality over the codebase. Like -
change orginal function `addDatabankEntries(dbname, dict_entries, verbose=True)` to something like -> `addDatabankEntries_configformat(dbname, dict_entries, verbose=True)`
and
`addDatabankEntriesJSON(dbname, dict_entries, verbose=True)` ->  `addDatabankEntries(dbname, dict_entries, verbose=True)`
So rest of the code remains unchanged ?

**Comment by anandxkumar** [comment]
Yes that will be better. And the original functions will  be replaced by the JSON functionality. 
Also for adding the test to check whether file is getting converted or not what about the following idea:

we can add a test in `test_config.py,` we will pass a flag parameter in the convert function and while calling the test we first use `setup_line_database_configformat` with the flag that will create a test config file suppose `.radistest `and then convert it to save as `radistest.json`, then we use the `setup_line_database` to create `radistest2.json` , and load both JSONs in variables and then delete the files we created (So nothing additional is being added) and check whether both variables are similar or not.
Is this approach appropriate to test the convert function or should we scrap the test idea?

**Comment by anandxkumar** [comment]
Hello @erwanp , I have replaced all the original functions with JSON ones. Now all the info will be stored in `~/radis.json` instead of `~/.radis`, it also passed all the test cases.  
One last thing we need to do is to automate the convertToRadisJSON() function but I'm not sure how will we automate, like will it be executed when someone wants to calculates a spectrum so that we will always check before executing the functions like `eq_spectrum` , `calc_spectrum`, etc?

File: radis/misc/__init__.py [hunk 4]
```diff
@@ -48,3 +48,9 @@
 from .progress_bar import ProgressBar
 from .signal import resample, resample_even
 from .utils import DatabankNotFound, NotInstalled, getProjectRoot
+
+# Checking `~radis.json` exist or not, if not then converting `~.radis` into `~/radis.json`
+try:
+    automatic_conversion()
+except:
+    raise ("Couldn't Convert `.radis` to `radis.json`")
```
Comment (erwanp) [inline radis/misc/__init__.py:56]: keep the inheritance of errors in the error log : 
```suggestion
except Exception as err:
    raise ValueError("Couldn't Convert `.radis` to `radis.json`") from err
```

File: radis/misc/__init__.py [hunk 5]
```diff
@@ -48,3 +48,9 @@
 from .progress_bar import ProgressBar
 from .signal import resample, resample_even
 from .utils import DatabankNotFound, NotInstalled, getProjectRoot
+
+# Checking `~radis.json` exist or not, if not then converting `~.radis` into `~/radis.json`
```
Comment (erwanp) [inline radis/misc/__init__.py:52]: What about moving this in [config.py/getConfig](https://github.com/radis/radis/blob/7a5b9b3179c6fc75eea626e98cf3e3e0810ff46c/radis/misc/config.py#L51) , if jsonfile is not found ?

**Review by erwanp — CHANGES_REQUESTED** [review]

**Comment by erwanp** [comment]
Review sent, we're almost there. Maybe one scenario you didn't see : 
- if .radis has been converted to .json we want users to stop using/editing the .radis afterwards. Could add a DeprecatedWarning in getConfig if json file is found, and .radis is also found, to explain that the config file was replaced in 0.9.29 and ask the user to remove it ?

File: radis/misc/__init__.py (hunk 5 above)
Comment (anandxkumar) [reply radis/misc/__init__.py:52]: Yes, definitely a better approach!

**Comment by anandxkumar** [comment]
Hi @erwanp , I shifted the `automatic_conversion` function into the `getConfig` function in the `config.py.` I'm using `warnings.warn` to give the Deprecation issue, because other approaches like raise DeprecationWarning and warnings.simple_filter would terminate the executing code and raise assertion error and I believe we don't want to do that.

File: radis/misc/config.py [hunk 6]
```diff
@@ -278,11 +279,110 @@ def getConfig_configformat():
...
+    """Checks whether `~/radis.json` exists.
+    If not then we use
+    :func:`~radis.misc.config.convertRadisToJSON`
+    to convert `~/.radis` into `~/radis.json`
+    """
+
+    # If `~/.radis` and `~/radis.json` both found, raises DeprecationWarning
+    if exists(CONFIG_PATH_JSON) and exists(CONFIG_PATH):
+        warnings.warn(
+            DeprecationWarning(
```
Comment (erwanp) [inline radis/misc/config.py:359]: 👍

**Review by erwanp — APPROVED** [review]

**Comment by erwanp** [comment]
Hello !! All good for the code, so the docs are the only thing missing. 

You can probably go for the easiest : 
- find/replace every occurence of ~/.radis in the docstrings by ~/radis.json
- in [lbl.rst](https://github.com/radis/radis/blob/develop/docs/lbl/lbl.rst#configuration-file), change the config sections by a typical json.  (maybe the easiest is to use your convert function ?)

Btw, Json can be quite unreadable without line returns. Does the converter add line returns ?

**Comment by anandxkumar** [comment]
Hi @erwanp, yes the JSON file is properly indented and stores all data line-wise in the JSON file, but the `getConfig` can be improved to get data line-wise like right now it's like the below when we load data from  `getConfig`
```
{'database': {'MY-HITEMP-CO2': {'path': ['D:\\Databases\\HITEMP-CO2\\hitemp_07',
    'D:\\Databases\\HITEMP-CO2\\hitemp_08',
    'D:\\Databases\\HITEMP-CO2\\hitemp_09'],
   'parfuncfmt': 'hapi'}}}
```
I can solve it. Also, I didn't account for the comments in the config file  -
```
[MY-HITEMP-CO2]                  #  your databank name: use this in calc_spectrum()
                                 #  or SpectrumFactory.load_databank()
path =                           #  no "", multipath allowed
       D:\Databases\HITEMP-CO2\hitemp_07
       D:\Databases\HITEMP-CO2\hitemp_08
       D:\Databases\HITEMP-CO2\hitemp_09
format = hitran                  #  'hitran' (HITRAN/HITEMP), 'cdsd-hitemp', 'cdsd-4000'
                                 # databank text file format. More info in
                                 # SpectrumFactory.load_databank function.
parfuncfmt = hapi                # calculate partition functions
```
So while converting to JSON unnecessary data will be also be converted, should I take care of this situation also or we can leave it because no one would edit (add comments) the configuration file anyway?

**Comment by erwanp** [comment]
Good point here : JSON doesn't accept comments, but we may want to let users have them in their config files. 

Exemple we could have : 
```javascript
{'database':   {
     'MY-HITEMP-CO2':  {                      //  your databank name: use this in calc_spectrum()
                                                            //  or SpectrumFactory.load_databank()
           'path':  [
                  'D:\\Databases\\HITEMP-CO2\\hitemp_07',
                  'D:\\Databases\\HITEMP-CO2\\hitemp_08',
                  'D:\\Databases\\HITEMP-CO2\\hitemp_09'],
         'format': 'hitran',                        // 'hitran' (HITRAN/HITEMP), 'cdsd-hitemp', 'cdsd-4000'
                                                           //  databank text file format. More info in
                                                           // SpectrumFactory.load_databank function.
          'parfuncfmt': 'hapi'                   //  calculate partition functions
                      }
          }
}
```

One solution I can think of is to use `json_minify` to clean the json and discard the comments before loading it : 

```
json.loads(json_minify(in_string))
```
https://github.com/getify/JSON.minify/tree/python

If you like it, we can add it to the [environment.yml](https://github.com/radis/radis/blob/develop/environment.yml)


-----

Btw, the "format" key was missing in your example above, maybe it's just a typo ? . 

---

**Comment by anandxkumar** [comment]
Yes, the missing 'format' key was a typo when I was removing the comments.
The comment thing wasn't implemented before also. Like if we add comments in .radis file like this for eg
```
[MY-HITEMP-CO2]                  #  your databank name: use this in calc_spectrum()
                                 #  or SpectrumFactory.load_databank()
path =                           #  no "", multipath allowed
       D:\Databases\HITEMP-CO2\hitemp_07
       D:\Databases\HITEMP-CO2\hitemp_08
       D:\Databases\HITEMP-CO2\hitemp_09
format = hitran                  #  'hitran' (HITRAN/HITEMP), 'cdsd-hitemp', 'cdsd-4000'
                                 # databank text file format. More info in
                                 # SpectrumFactory.load_databank function.
parfuncfmt = hapi                # calculate partition functions
```
And get the databank entries `getDatabankEntries_configformat`, the data is in this form -
```
{'path': ['#  no "", multipath allowed',
  'D:\\Databases\\HITEMP-CO2\\hitemp_07',
  'D:\\Databases\\HITEMP-CO2\\hitemp_08',
  'D:\\Databases\\HITEMP-CO2\\hitemp_09'],
 'format': "hitran                  #  'hitran' (HITRAN/HITEMP), 'cdsd-hitemp', 'cdsd-4000'",
 'parfuncfmt': 'hapi                # calculate partition functions'}
```
So beforehand this thing wasn't meant to be done.
On the other hand, we can definitely implement the comment thing in JSON file,
From the below piece of code, we can implement the same

```
import json_minify

file_name = CONFIG_PATH_JSON
file_data = open(file_name, "r", 1).read() 
s = json_minify.json_minify(file_data)
json_data = json.loads(s)

```
eg -
```
{
   "database": {
      "MY-HITEMP-CO2": {			//  your databank name: use this in calc_spectrum()	
         "path": [
            "D:\\Databases\\HITEMP-CO2\\hitemp_07", //Added database path examples
            "D:\\Databases\\HITEMP-CO2\\hitemp_08",
            "D:\\Databases\\HITEMP-CO2\\hitemp_09"
         ],
         "format": "hitran",
         "parfuncfmt": "hapi"
      }
   }
}
```
-> 
```
{'database': {'MY-HITEMP-CO2': {'path': ['D:\\Databases\\HITEMP-CO2\\hitemp_07',
    'D:\\Databases\\HITEMP-CO2\\hitemp_08',
    'D:\\Databases\\HITEMP-CO2\\hitemp_09'],
   'format': 'hitran',
   'parfuncfmt': 'hapi'}}}

```
So a good addition in my opinion.

**Comment by erwanp** [comment]
@anandxkumar if you agree we'll merge this, and we can improve the related docs (and ability to add Comments) in the later release, i.e. 0.9.30. May also be useful to fix problems encountered by users in the meantime !

**Comment by anandxkumar** [comment]
Ok @erwanp, fine with me! I will work on implementing the features and docs discussed later :) . Yes a good idea indeed cause there will be some flaws with the implementation so it will be good to see how it works on everyone's system :) and fix any related bugs!

**Pull request merged**